### DISCUSSION TITLE
Reset Database / Purge tables

### DISCUSSION THREAD

**Discussion by Arslan007** [Azure/azure-mobile-apps#930] [Q&A]
In our app we have a feature where user can reset his data and for that we are purging the tables but there are some internal tables like __config, __operations & __errors how we can purge them. Is there any convenient method to cleanup/purge them or we have to use same approach like the other normal tables.

**Comment by adrianhall** [comment]
If you are resetting the data, just delete the database file (after closing it), then re-initialize the client.  

If you don't want to do that, then the Sqlite database is the state file, but the __config is cached in memory, so you can remove all the entries in __operations and __errors using the [ExecuteNonQueryInternal()](https://github.com/Azure/azure-mobile-apps/blob/main/sdk/dotnet/src/Microsoft.Datasync.Client.SQLiteStore/OfflineSQLiteStore.cs#L468) method - something like

```csharp
_ = await store.ExecuteNonQueryInternal("DELETE FROM __operations;");
```

However, this method is protected, so you will have to sub-class it and potentially add your "reset" method as a new method to that sub-class.  Something like:

```csharp
public class MyOfflineStore(string connectionString) : OfflineSqliteStore(connectionString)
{
  private Lazy<IDeltaTokenStore> _dts = new(() => new SQLiteDeltaTokenStore(this));

  public override Task<IDeltaTokenStore> GetDeltaTokenStoreAsync(CancellationToken ct = default)
    => _dts.Value;

  public async Task ResetDatabase(CancellationToken ct= default) 
  {
    List<string> tables = await GetTablesAsync(ct);
    IDeltaTokenStore dts = await GetDeltaTokenStoreAsync(ct);
    foreach (string table in tables) 
    {
        await ExecuteNonQueryInternal($"DELETE FROM {table};");
        var queryIds = (await dts.GetDeltaTokenQueryIdsForTableAsync(table, ct)).ToList();
        foreach (var queryId in queryIds)
        {
          await dts.ResetDeltaTokenAsync(table, queryId, ct);
        }
    }
    await ExecuteNonQueryInternal("DELETE FROM __operations;");
    await ExecuteNonQueryInternal("DELETE FROM __errors;");
}
```

The delta token store implementation in the SQLiteStore is not the best, but it's been revamped in the new Community Toolkit version, so I'm not worried about it in Azure Mobile Apps.
